File: mobile/flutter/lib/features/lending/lending_service.dart
// Lending Service - Flutter Mobile
// Complete lending/borrowing with real backend

import 'dart:convert';
import 'package:http/http.dart' as http;

class LendingService {
  static const String API_BASE = 'http://localhost:8443/api/v1';
  String? _token;
  
  LendingService({String? token}) : _token = token;
  
  Map<String, String> get _headers => {
    'Content-Type': 'application/json',
    if (_token != null) 'Authorization': 'Bearer $_token',
  };
  
  // Get all lending pools
  Future<List<LendingPool>> getPools() async {
    final response = await http.get(
      Uri.parse('$API_BASE/lending/pools'),
      headers: _headers,
    );
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return (data['data'] as List).map((p) => LendingPool.fromJson(p)).toList();
    }
    return [];
  }
  
  // Get pool data for specific token
  Future<LendingPool> getPool(String token) async {
    final response = await http.get(
      Uri.parse('$API_BASE/lending/pools/$token'),
      headers: _headers,
    );
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return LendingPool.fromJson(data['data']);
    }
    throw Exception('Failed to load pool');
  }
  
  // Supply assets to pool
  Future<LendingPosition> supply(String token, double amount) async {
    final response = await http.post(
      Uri.parse('$API_BASE/lending/supply'),
      headers: _headers,
      body: json.encode({
        'token': token,
        'amount': amount,
      }),
    );
    
    if (response.statusCode == 201) {
      final data = json.decode(response.body);
      return LendingPosition.fromJson(data['data']);
    }
    throw Exception('Failed to supply');
  }
  
  // Borrow from pool
  Future<LendingPosition> borrow(String token, double amount) async {
    final response = await http.post(
      Uri.parse('$API_BASE/lending/borrow'),
      headers: _headers,
      body: json.encode({
        'token': token,
        'amount': amount,
      }),
    );
    
    if (response.statusCode == 201) {
      final data = json.decode(response.body);
      return LendingPosition.fromJson(data['data']);
    }
    throw Exception('Failed to borrow');
  }
  
  // Repay borrowed amount
  Future<bool> repay(String token, double amount) async {
    final response = await http.post(
      Uri.parse('$API_BASE/lending/repay'),
      headers: _headers,
      body: json.encode({
        'token': token,
        'amount': amount,
      }),
    );
    
    return response.statusCode == 200;
  }
  
  // Withdraw supplied assets
  Future<bool> withdraw(String token, double amount) async {
    final response = await http.post(
      Uri.parse('$API_BASE/lending/withdraw'),
      headers: _headers,
      body: json.encode({
        'token': token,
        'amount': amount,
      }),
    );
    
    return response.statusCode == 200;
  }
  
  // Get user's positions
  Future<List<LendingPosition>> getUserPositions() async {
    final response = await http.get(
      Uri.parse('$API_BASE/lending/positions'),
      headers: _headers,
    );
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return (data['data'] as List).map((p) => LendingPosition.fromJson(p)).toList();
    }
    return [];
  }
  
  // Get health factor
  Future<double> getHealthFactor() async {
    final response = await http.get(
      Uri.parse('$API_BASE/lending/health'),
      headers: _headers,
    );
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return (data['data'] ?? 0).toDouble();
    }
    return 999.0;
  }
  
  // Liquidate position
  Future<bool> liquidate(String userId, String token) async {
    final response = await http.post(
      Uri.parse('$API_BASE/lending/liquidate'),
      headers: _headers,
      body: json.encode({
        'userId': userId,
        'token': token,
      }),
    );
    
    return response.statusCode == 200;
  }
}

class LendingPool {
  final String token;
  final String name;
  final String symbol;
  final String icon;
  final double totalSupplied;
  final double totalBorrowed;
  final double supplyAPY;
  final double borrowAPY;
  final double liquidity;
  final DateTime updatedAt;
  
  LendingPool({
    required this.token,
    required this.name,
    required this.symbol,
    required this.icon,
    required this.totalSupplied,
    required this.totalBorrowed,
    required this.supplyAPY,
    required this.borrowAPY,
    required this.liquidity,
    required this.updatedAt,
  });
  
  factory LendingPool.fromJson(Map<String, dynamic> json) {
    return LendingPool(
      token: json['token'] ?? '',
      name: json['name'] ?? '',
      symbol: json['symbol'] ?? '',
      icon: json['icon'] ?? '',
      totalSupplied: (json['totalSupplied'] ?? 0).toDouble(),
      totalBorrowed: (json['totalBorrowed'] ?? 0).toDouble(),
      supplyAPY: (json['supplyAPY'] ?? 0).toDouble(),
      borrowAPY: (json['borrowAPY'] ?? 0).toDouble(),
      liquidity: (json['liquidity'] ?? 0).toDouble(),
      updatedAt: DateTime.parse(json['updatedAt']),
    );
  }
}

class LendingPosition {
  final String id;
  final String token;
  final double supplied;
  final double borrowed;
  final double apy;
  final double accumulated;
  final String status;
  final DateTime suppliedAt;
  
  LendingPosition({
    required this.id,
    required this.token,
    required this.supplied,
    required this.borrowed,
    required this.apy,
    required this.accumulated,
    required this.status,
    required this.suppliedAt,
  });
  
  factory LendingPosition.fromJson(Map<String, dynamic> json) {
    return LendingPosition(
      id: json['id'] ?? '',
      token: json['token'] ?? '',
      supplied: (json['supplied'] ?? 0).toDouble(),
      borrowed: (json['borrowed'] ?? 0).toDouble(),
      apy: (json['apy'] ?? 0).toDouble(),
      accumulated: (json['accumulated'] ?? 0).toDouble(),
      status: json['status'] ?? 'ACTIVE',
      suppliedAt: DateTime.parse(json['suppliedAt']),
    );
  }
}

File: mobile/ios/Runner/Services/LendingService.swift
import Foundation

/**
 * Lending Service - iOS Native Implementation
 * Real backend connection to Go lending service
 */
class LendingService {
    static let API_BASE = "http://localhost:8443/api/v1"
    var authToken: String?
    
    init(token: String? = nil) {
        self.authToken = token
    }
    
    private func createRequest(endpoint: String, method: String = "GET", body: [String: Any]? = nil) -> URLRequest? {
        guard let url = URL(string: "\(Self.API_BASE)\(endpoint)") else { return nil }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token = authToken {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        if let body = body {
            request.httpBody = try? JSONSerialization.data(withJSONObject: body)
        }
        return request
    }
    
    /// Get all lending pools
    func getPools(completion: @escaping (Result<[LendingPool], Error>) -> Void) {
        guard let request = createRequest(endpoint: "/lending/pools") else {
            completion(.failure(LendingError.invalidRequest))
            return
        }
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let poolsData = json["data"] as? [[String: Any]] else {
                completion(.failure(LendingError.invalidResponse))
                return
            }
            
            let pools = poolsData.compactMap { poolJson -> LendingPool? in
                guard let token = poolJson["token"] as? String,
                      let name = poolJson["name"] as? String else { return nil }
                return LendingPool(
                    token: token,
                    name: name,
                    totalSupplied: poolJson["totalSupplied"] as? Double ?? 0,
                    totalBorrowed: poolJson["totalBorrowed"] as? Double ?? 0,
                    supplyAPY: poolJson["supplyAPY"] as? Double ?? 0,
                    borrowAPY: poolJson["borrowAPY"] as? Double ?? 0,
                    liquidity: poolJson["liquidity"] as? Double ?? 0
                )
            }
            completion(.success(pools))
        }.resume()
    }
    
    /// Supply assets to pool
    func supply(token: String, amount: Double, completion: @escaping (Result<LendingPosition, Error>) -> Void) {
        guard let request = createRequest(endpoint: "/lending/supply", method: "POST", body: [
            "token": token,
            "amount": amount
        ]) else {
            completion(.failure(LendingError.invalidRequest))
            return
        }
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let positionData = json["data"] as? [String: Any] else {
                completion(.failure(LendingError.invalidResponse))
                return
            }
            
            let position = LendingPosition(
                id: positionData["id"] as? String ?? "",
                token: positionData["token"] as? String ?? "",
                supplied: positionData["supplied"] as? Double ?? 0,
                borrowed: positionData["borrowed"] as? Double ?? 0,
                apy: positionData["apy"] as? Double ?? 0,
                status: positionData["status"] as? String ?? ""
            )
            completion(.success(position))
        }.resume()
    }
    
    /// Borrow from pool
    func borrow(token: String, amount: Double, completion: @escaping (Result<LendingPosition, Error>) -> Void) {
        guard let request = createRequest(endpoint: "/lending/borrow", method: "POST", body: [
            "token": token,
            "amount": amount
        ]) else {
            completion(.failure(LendingError.invalidRequest))
            return
        }
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let positionData = json["data"] as? [String: Any] else {
                completion(.failure(LendingError.invalidResponse))
                return
            }
            
            let position = LendingPosition(
                id: positionData["id"] as? String ?? "",
                token: positionData["token"] as? String ?? "",
                supplied: positionData["supplied"] as? Double ?? 0,
                borrowed: positionData["borrowed"] as? Double ?? 0,
                apy: positionData["apy"] as? Double ?? 0,
                status: positionData["status"] as? String ?? ""
            )
            completion(.success(position))
        }.resume()
    }
    
    /// Repay borrowed amount
    func repay(token: String, amount: Double, completion: @escaping (Result<Bool, Error>) -> Void) {
        guard let request = createRequest(endpoint: "/lending/repay", method: "POST", body: [
            "token": token,
            "amount": amount
        ]) else {
            completion(.failure(LendingError.invalidRequest))
            return
        }
        
        URLSession.shared.dataTask(with: request) { _, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            if let httpResponse = response as? HTTPURLResponse {
                completion(.success(httpResponse.statusCode == 200))
            } else {
                completion(.failure(LendingError.invalidResponse))
            }
        }.resume()
    }
    
    /// Withdraw supplied assets
    func withdraw(token: String, amount: Double, completion: @escaping (Result<Bool, Error>) -> Void) {
        guard let request = createRequest(endpoint: "/lending/withdraw", method: "POST", body: [
            "token": token,
            "amount": amount
        ]) else {
            completion(.failure(LendingError.invalidRequest))
            return
        }
        
        URLSession.shared.dataTask(with: request) { _, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            if let httpResponse = response as? HTTPURLResponse {
                completion(.success(httpResponse.statusCode == 200))
            } else {
                completion(.failure(LendingError.invalidResponse))
            }
        }.resume()
    }
    
    /// Get user's positions
    func getUserPositions(completion: @escaping (Result<[LendingPosition], Error>) -> Void) {
        guard let request = createRequest(endpoint: "/lending/positions") else {
            completion(.failure(LendingError.invalidRequest))
            return
        }
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let positionsData = json["data"] as? [[String: Any]] else {
                completion(.failure(LendingError.invalidResponse))
                return
            }
            
            let positions = positionsData.compactMap { posJson -> LendingPosition? in
                guard let id = posJson["id"] as? String,
                      let token = posJson["token"] as? String else { return nil }
                return LendingPosition(
                    id: id,
                    token: token,
                    supplied: posJson["supplied"] as? Double ?? 0,
                    borrowed: posJson["borrowed"] as? Double ?? 0,
                    apy: posJson["apy"] as? Double ?? 0,
                    status: posJson["status"] as? String ?? ""
                )
            }
            completion(.success(positions))
        }.resume()
    }
    
    /// Get health factor
    func getHealthFactor(completion: @escaping (Result<Double, Error>) -> Void) {
        guard let request = createRequest(endpoint: "/lending/health") else {
            completion(.failure(LendingError.invalidRequest))
            return
        }
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let healthFactor = json["data"] as? Double else {
                completion(.failure(LendingError.invalidResponse))
                return
            }
            completion(.success(healthFactor))
        }.resume()
    }
}

// MARK: - Data Models
struct LendingPool {
    let token: String
    let name: String
    let totalSupplied: Double
    let totalBorrowed: Double
    let supplyAPY: Double
    let borrowAPY: Double
    let liquidity: Double
}

struct LendingPosition {
    let id: String
    let token: String
    let supplied: Double
    let borrowed: Double
    let apy: Double
    let status: String
}

enum LendingError: Error {
    case invalidRequest
    case invalidResponse
}
